metabase/metabase
# Add referenced question columns to autocomplete results

This PR extends autocomplete in the native SQL editor to include columns from saved questions and models that are referenced with template tags in the query.

[Notion block link](https://www.notion.so/metabase/Improve-native-query-UX-on-models-5f4505503cc64b4fa0b15e1d37651503#ebc8cac2c8364036a7b35bf91ae56ac5)

![image](https://user-images.githubusercontent.com/39073188/186956116-a1f00207-0ca0-4366-9cfb-c2257c147577.png)

I originally implemented this purely on the backend [here](https://github.com/metabase/metabase/pull/25025), but switched to implement it only on the FE to save unnecessary DB queries.

### **How to verify**

1. Create a saved question from a native SQL query, with a column name that is unique so you can search for it later, e.g. `a_column`.
2. Create another native SQL query, and with a template tag referencing that question, e.g. `select * from {{#<question-id>}}`
3. Start typing the column name from step 1 into the editor, and it should appear in the autocomplete popup. (test 1)
4. Save the question and refresh the page. Open the SQL editor and type start typing the column name. It should appear again. (test 2)

### BE changes 

None.

### FE changes

I've updated the `completer` function for the `aceLanguageTools` class, which runs when the user types. The completer function previously returned table fields and models, but now it gets all the column names from questions referenced in the native query. These column names will appear in the autocomplete popover as the use types.
